pkg/engine/planner/physical/optimizer.go

```diff
@@ -261,6 +269,144 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
 
 var _ rule = (*projectionPushdown)(nil)
 
+// parseKeysPushdown determines which keys need to be parsed based on downstream operations
+type parseKeysPushdown struct {
+	plan *Plan
+}
```

Comment:
The keys to parse acts the same as a projection, doesn't it? Maybe we could wrap this into `applyProjectionPushdown` to avoid adding a new stage?

That would also help to get remove of all the `extract*` helper functions since you could just use the existing `projections` parameter in that function.